etl/extractor/platform_impl.go

```diff
@@ -33,220 +30,160 @@ type PlatformExtractor struct {
-				}
+		pu, err = e.pulseExtractor.GetNextFinalizedPulse(ctx, int64(before.PulseNumber))
+		if err != nil { // network error ?
+			if strings.Contains(err.Error(), pulse.ErrNotFound.Error()) { // seems this pulse already last
+				time.Sleep(halfPulse)
+				continue
 			}
+			log.Errorf("retrievePulses(): before=%d err=%s", before, err)
+			pu = &before
+			time.Sleep(time.Second)
+			continue
 		}
-	}()
-}
-
-func (m *PlatformExtractor) getJetDropsContinuously(ctx context.Context) {
-	logger := belogger.FromContext(ctx)
-	var pulse uint32 = 0
-	var err error
-
-	// try to get current pulse with attempts
-	for i := 0; i < m.pulseExtractAttempts; i++ {
-		pulse, err = m.pulseExtractor.GetCurrentPulse(ctx)
-		if err != nil {
-			logger.Warnf("trying to get current pulse, attempt: %d", i)
-			time.Sleep(time.Duration(pulseDelta) * time.Second)
-		} else {
-			break
+		if pu.PulseNumber == before.PulseNumber { // no new pulse happens
+			time.Sleep(halfPulse)
+			continue
 		}
-	}
 
-	// fatal and exit if could not get current pulse
-	if pulse == 0 || err != nil {
-		logger.Fatalf("could not get current pulse number after %d attempts", m.pulseExtractAttempts)
-	}
-
-	logger.Infof("current pulse number: %d", pulse)
-	m.request.PulseNumber = insolar.PulseNumber(pulse)
-	m.request.RecordNumber = 0
-	m.getJetDrops(ctx, m.request, 0, math.MaxUint32, false)
-}
+		log.Debug("retrievePulses(): Done")
 
-func logGRPCError(ctx context.Context, err error) {
-	log := belogger.FromContext(ctx)
-	if err != nil {
-		log.Debug("Data request failed: ", err)
-		log.Error(errors.Wrapf(err, "failed to get gRPC stream from exporter.Export method").Error())
-	}
-}
+		go e.retrieveRecords(ctx, pu)
 
-func logIfErrorReceived(ctx context.Context, err error, request interface{}) {
-	if err != nil {
-		belogger.FromContext(ctx).Debug("received error value from records gRPC stream %v", request)
-	}
-}
+		if until <= 0 { // we are going on the edge of history
+			time.Sleep(halfPulse)
+		} else if pu.PulseNumber >= insolar.PulseNumber(until) { // we are at the end
+			return
+		}
 
-func isEOF(ctx context.Context, err error) bool {
-	if err == io.EOF {
-		belogger.FromContext(ctx).Debug("EOF received, quit")
-		return true
 	}
-	return false
 }
 
-func closeStream(ctx context.Context, stream exporter.RecordExporter_ExportClient) {
-	if stream != nil {
-		streamError := stream.CloseSend()
-		if streamError != nil {
-			belogger.FromContext(ctx).Warn("Error closing stream: ", streamError)
+// retrieveRecords - retrieves all records for specified pulse and puts this to channel
+func (e *PlatformExtractor) retrieveRecords(ctx context.Context, pu *exporter.FullPulse) {
+	logger := belogger.FromContext(ctx)
+	log := logger.WithField("pulse_number", pu.PulseNumber)
+	log.Debug("retrieveRecords(): Start")
+	jetDrops := &types.PlatformJetDrops{Pulse: pu} // save pulse info
+
+	for { // each portion
+		stream, err := e.client.Export(ctx, &exporter.GetRecords{PulseNumber: pu.PulseNumber,
```

Comment:
Shouldn't we check `<-ctx.Done()` before this too?

Reply:
No, we rapidly falling back to each record and catching it there.

Reply:
And if error happened and we don't iterate over records?

Reply:
ok, it isn't a problem